File: test/rs-stub/tests/map_key_malformed_test.rs
// PR-F (M24) — cross-language malformed map-key error consistency.
//
// Verifies that `Holder::from_json` surfaces a "malformed key: ..." error when the
// JSON map key cannot be parsed back into the id type. The explicit JSON decoder routes
// the parse error through `BaboonCodecError::decoder_failure(format!("malformed key: {}", e))`,
// so the resulting error message starts with "malformed key".
//
// Uses the my.ok.m19.singleid fixture (id ItemId { v: uid }; root data Holder { m: map[ItemId, str] }).
// Generated symbols are produced by mdl :test-gen-regular-adt under target/test-regular/rs-stub/.

use baboon_rs_stub::my::ok::m19::singleid::holder::Holder;

#[test]
fn holder_json_decode_returns_malformed_key_for_bad_map_key() {
    let bad_json = r#"{"m":{"not_a_valid_id":"v"}}"#;
    let err = Holder::from_json(bad_json).expect_err("expected a decode error for malformed map key");
    let msg = match &err {
        baboon_rs_stub::any_opaque::BaboonCodecError::DecoderFailure { message, .. } => message.clone(),
        other => panic!("expected a DecoderFailure but got: {:?}", other),
    };
    assert!(msg.starts_with("malformed key: "),
        "expected message to start with 'malformed key: ' but got: {msg}");
}

File: test/kt-stub-kmp/src/test/kotlin/runtime/MapKeyMalformedTest.kt
// PR-F (M24) — cross-language malformed map-key error consistency.
//
// Verifies that decoding a JSON object whose map-key cannot be parsed back into the
// id type throws BaboonCodecException.DecoderFailure with message containing
// "malformed key". Replaces the prior unchecked `as Either.Right` cast.
//
// Mirrored from test/kt-stub/src/test/kotlin/runtime/MapKeyMalformedTest.kt for
// PR-25.5a (M25 KMP runtime test parity). No KMP-specific adaptation needed:
// the test uses only kotlinx.serialization (multiplatform-friendly) and generated
// Holder_JsonCodec. Uses the my.ok.m19.singleid fixture; generated symbols are
// produced by mdl :test-gen-regular-adt under target/test-regular/kt-stub-kmp/.
package runtime

import baboon.runtime.shared.BaboonCodecContext
import baboon.runtime.shared.BaboonCodecException
import kotlinx.serialization.json.Json
import my.ok.m19.singleid.Holder_JsonCodec
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class MapKeyMalformedTest {

    private val ctx = BaboonCodecContext.Compact

    @Test
    fun holderJsonDecodeThrowsDecoderFailureForMalformedMapKey() {
        val badJson = """{"m":{"not_a_valid_id":"v"}}"""
        val node = Json.parseToJsonElement(badJson)
        val ex = assertThrows(BaboonCodecException.DecoderFailure::class.java) {
            Holder_JsonCodec.decode(ctx, node)
        }
        assertTrue(
            ex.message?.contains("malformed key") == true,
            "unexpected message: ${ex.message}"
        )
    }
}
